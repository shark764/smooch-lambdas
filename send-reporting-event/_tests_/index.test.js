const uuidv1 = require('uuid/v4');

jest.mock('uuid/v4');

global.Date.prototype.toISOString = jest.fn(() => 'January 1 1970');

uuidv1.mockImplementation(() => '7534c040-534d-11ea-8aa0-c32d6a748e46');

global.process.env = {
  AWS_REGION: 'us-east-1',
  ENVIRONMENT: 'dev',
  ACCOUNT_ID: '667802d8-2260-436c-958a-2ee0f71f73f7',
};

const mockPublish = jest.fn()
  .mockImplementation(() => ({
    promise: () => ({}),
  }));

jest.mock('aws-sdk', () => ({
  SNS: jest.fn().mockImplementation(() => ({ publish: mockPublish })),
}));

const event = {
  Records: [{
    body: JSON.stringify({
      tenantId: '250faddb-9723-403a-9bd5-3ca710cb26e5',
      interactionId: '667802d8-2260-436c-958a-2ee0f71f73f0',
      topic: 'agent-message',
      appName: 'smooch',
      resourceId: '667802d8-2260-436c-958a-2ee0f71f73f2',
    }),
  }],
};

const { handler } = require('../index');

describe('send-reporting-event', () => {
  describe('Everthing is successful', () => {
    it('when the provided topic is customer-message', async () => {
      const mockEvent = {
        Records: [{
          body: JSON.stringify({
            tenantId: '250faddb-9723-403a-9bd5-3ca710cb26e5',
            interactionId: '667802d8-2260-436c-958a-2ee0f71f73f0',
            topic: 'customer-message',
            appName: 'smooch',
            resourceId: '667802d8-2260-436c-958a-2ee0f71f73f2',
          }),
        }],
      };
      await handler(mockEvent);
      expect(mockPublish.mock.calls[0]).toMatchSnapshot();
    });

    describe('Walkthrough', () => {
      beforeAll(async () => {
        jest.clearAllMocks();
        await handler(event);
      });
      it('passes in the correct arguments to sns.publish()', async () => {
        expect(mockPublish.mock.calls).toMatchSnapshot();
      });
    });
  });
  it('throws an error when there is a problem sending reporting event', async () => {
    try {
      mockPublish.mockRejectedValueOnce(new Error());
      await handler(event);
    } catch (error) {
      expect(Promise.reject(new Error('Error Sending Reporting Event'))).rejects.toThrowErrorMatchingSnapshot();
    }
  });

  it('throws an error when the topic received are not supported', async () => {
    try {
      const mockEvent = {
        Records: [{
          body: JSON.stringify({
            tenantId: '250faddb-9723-403a-9bd5-3ca710cb26e5',
            interactionId: '667802d8-2260-436c-958a-2ee0f71f73f0',
            topic: 'topic',
            appName: 'smooch',
            resourceId: '667802d8-2260-436c-958a-2ee0f71f73f2',
          }),
        }],
      };
      await handler(mockEvent);
    } catch (error) {
      expect(error.message).toBe('Topic received not supported');
    }
  });
});
