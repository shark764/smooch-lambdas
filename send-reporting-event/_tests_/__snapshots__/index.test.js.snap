// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`send-reporting-event Everthing is successful Walkthrough passes in the correct arguments to sns.publish() 1`] = `
Array [
  Array [
    Object {
      "Message": "dist-event",
      "MessageAttributes": Object {
        "app-id": Object {
          "DataType": "String",
          "StringValue": "55448dde-5fa1-416f-a55a-19537cc63c94",
        },
        "app-name": Object {
          "DataType": "String",
          "StringValue": "smooch",
        },
        "encoding": Object {
          "DataType": "String",
          "StringValue": "application/json",
        },
        "message": Object {
          "DataType": "String",
          "StringValue": "{\\"psychopomp/version\\":\\"psychopomp.messages.reporting/agent-message\\",\\"psychopomp/type\\":\\"agent-message\\",\\"event-id\\":\\"7534c040-534d-11ea-8aa0-c32d6a748e46\\",\\"timestamp\\":\\"January 1 1970Z\\",\\"app-name\\":\\"smooch\\",\\"app-id\\":\\"55448dde-5fa1-416f-a55a-19537cc63c94\\",\\"tenant-id\\":\\"250faddb-9723-403a-9bd5-3ca710cb26e5\\",\\"interaction-id\\":\\"667802d8-2260-436c-958a-2ee0f71f73f0\\",\\"agent-id\\":\\"667802d8-2260-436c-958a-2ee0f71f73f2\\",\\"message-id\\":\\"7534c040-534d-11ea-8aa0-c32d6a748e46\\"}",
        },
        "topic-key": Object {
          "DataType": "String",
          "StringValue": "agent-message",
        },
      },
      "TopicArn": "arn:aws:sns:us-east-1:667802d8-2260-436c-958a-2ee0f71f73f7:us-east-1-dev-agent-message",
    },
  ],
]
`;

exports[`send-reporting-event Everthing is successful when the provided topic is customer-message 1`] = `
Array [
  Object {
    "Message": "dist-event",
    "MessageAttributes": Object {
      "app-id": Object {
        "DataType": "String",
        "StringValue": "55448dde-5fa1-416f-a55a-19537cc63c94",
      },
      "app-name": Object {
        "DataType": "String",
        "StringValue": "smooch",
      },
      "encoding": Object {
        "DataType": "String",
        "StringValue": "application/json",
      },
      "message": Object {
        "DataType": "String",
        "StringValue": "{\\"psychopomp/version\\":\\"psychopomp.messages.reporting/customer-message\\",\\"psychopomp/type\\":\\"customer-message\\",\\"event-id\\":\\"7534c040-534d-11ea-8aa0-c32d6a748e46\\",\\"timestamp\\":\\"January 1 1970Z\\",\\"app-name\\":\\"smooch\\",\\"app-id\\":\\"55448dde-5fa1-416f-a55a-19537cc63c94\\",\\"tenant-id\\":\\"250faddb-9723-403a-9bd5-3ca710cb26e5\\",\\"interaction-id\\":\\"667802d8-2260-436c-958a-2ee0f71f73f0\\",\\"agent-id\\":\\"667802d8-2260-436c-958a-2ee0f71f73f2\\",\\"message-id\\":\\"7534c040-534d-11ea-8aa0-c32d6a748e46\\"}",
      },
      "topic-key": Object {
        "DataType": "String",
        "StringValue": "customer-message",
      },
    },
    "TopicArn": "arn:aws:sns:us-east-1:667802d8-2260-436c-958a-2ee0f71f73f7:us-east-1-dev-customer-message",
  },
]
`;

exports[`send-reporting-event throws an error when there is a problem sending reporting event 1`] = `"Error Sending Reporting Event"`;
